### PR TITLE
Add levenshtein_n to use without null terminated strings (and fix portability issues)

### PR DESCRIPTION
**Pull request by dcousens**
(title says it all)

### CODE REVIEW

**Comment by wooorm** [comment]
@dcousens Fancy! Could you also update the tests and docs though?

**Comment by wooorm** [comment]
@dcousens Ping!

**Comment by dcousens** [comment]
On my TODO :+1:

**Comment by dcousens** [comment]
@wooorm please see 5382ec1
Up to you if you want me to down-cast to an `unsigned int`...
It just means you need to cast in several places to make it work portably,  and then add checks if you're input length could surpass 32-bits (unlikely, but, whatever,  user choice,  I've done similar things...)

**Comment by dcousens** [comment]
@wooorm anything else required?

**Comment by dcousens** [comment]
@wooorm ping!

File: README.md [hunk 1]
```diff
@@ -35,6 +35,21 @@ main(int argc, char **argv) {
 }
```
Comment (wooorm) [inline README.md]: Why is this now `README.md`? I like `readme.md`!

File: README.md [hunk 2]
```diff
@@ -35,6 +35,21 @@ main(int argc, char **argv) {
...
 ```
 
+### `unsigned int levenshtein_n(const char *a, const size_t length, const char *b, const size_t bLength);`
```
Comment (wooorm) [inline README.md]: should this be `unsigned int`, or rather `size_t`?

File: test.c [hunk 3]
```diff
@@ -9,15 +9,18 @@ static int assertionCount = 0;
...
 
-  if (res != result) {
+  if (res != result || res_n != result) {
     errorCount++;
 
     fprintf(stderr, "\033[31m");
     fprintf(stderr,
-      "  (✖) For `%s` and `%s`. Expected `%d`, got `%d`",
+      "  (✖) For `%s` and `%s`. Expected `%zu`, got `%zu`",
       input, alternative, result, res
```
Comment (wooorm) [inline test.c]: should `res_n` also be printed? Or maybe use two assertions instead?

**Review by wooorm — CHANGES_REQUESTED** [review]
Argh, so sorry for the slow reply. Thanks for keeping on pinging me 👍

File: test.c (hunk 3 above)
Comment (dcousens) [reply test.c]: see latest

File: README.md (hunk 2 above)
Comment (dcousens) [reply README.md]: fixed

File: README.md (hunk 1 above)
Comment (dcousens) [reply README.md]: reverted (I used `README.md` as per GitHub standards,  but,  don't care tbh)

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
LGTM. Breaking change, right, so I’ll go with a semver major!

**Comment by wooorm** [comment]
If I  paste your new example into `example.c` and run `make run-example`, I get the following failures:

(after changing `%d` to `%zu`)

```
cc example.c -D_GNU_SOURCE -std=c99 -Wall -Wno-format-y2k -W -Wstrict-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wshadow -Wcast-align -Wbad-function-cast -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -c -o example.o
example.c:6:13: warning: multi-character character constant [-Wmultichar]
  char *a = 'foobar';
            ^
example.c:6:13: warning: character constant too long for its type
example.c:6:9: warning: incompatible integer to pointer conversion initializing 'char *' with an expression of type 'int' [-Wint-conversion]
  char *a = 'foobar';
        ^   ~~~~~~~~
example.c:7:13: warning: multi-character character constant [-Wmultichar]
  char *b = 'hello';
            ^
example.c:7:13: warning: character constant too long for its type
example.c:7:9: warning: incompatible integer to pointer conversion initializing 'char *' with an expression of type 'int' [-Wint-conversion]
  char *b = 'hello';
        ^   ~~~~~~~
6 warnings generated.
cc example.o levenshtein.o -o example
./example sitting kitten
make: *** [run-example] Segmentation fault: 11
```

...what do you suggest?

**Comment by dcousens** [comment]
~~@wooorm the warnings are because you are using `char*` for a `const char*`?~~
**edit**: strings in C need double quotes, e.g `"foobar"`.
See https://github.com/wooorm/levenshtein.c/pull/4

If I do
``` c
-  printf("%d\n", levenshtein(a, b));
+  printf("%zu\n", levenshtein(a, b));
```

``` bash
cc example.c -D_GNU_SOURCE -std=c99 -Wall -Wno-format-y2k -W -Wstrict-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wshadow -Wcast-align -Wbad-function-cast -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -c -o example.o
cc levenshtein.c -D_GNU_SOURCE -std=c99 -Wall -Wno-format-y2k -W -Wstrict-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch
-Wshadow -Wcast-align -Wbad-function-cast -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -c -o levenshtein.o
cc example.o levenshtein.o -o example
./example sitting kitten
3
```

**Pull request merged**